File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.py
class Solution:
    def evaluate(self, s, knowledge):
        mp = {}

        for t in knowledge:
            mp[t[0]] = t[1]

        ans = ""
        i = 0

        while i < len(s):
            if s[i] == '(':
                t = ""
                i += 1

                while i < len(s) and s[i] != ')':
                    t += s[i]
                    i += 1

                if t not in mp:
                    ans += '?'
                else:
                    ans += mp[t]

            else:
                ans += s[i]

            i += 1

        return ans

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.cpp
class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        unordered_map<string,string>mp;
        for(vector<string>&t:knowledge){
            mp[t[0]]=t[1];
        }
        string ans="";
        int i=0;
        while(i<s.size())
        {
            if(s[i]=='('){
                string t="";
                i++;
                while(i<s.size()&&s[i]!=')')
                {
                    t+=s[i];
                    i++;
                }
                if(!mp.count(t)){
                    ans+='?';
                }else{
                    ans+=mp[t];
                }
            }
            else{
                ans+=s[i];
            }
            i++;
        }
        return ans;
    }
};